config/crd/experimental/gateway.networking.k8s.io_httproutes.yaml

```
@@ -1692,6 +1692,21 @@ spec:
                         type: object
                       maxItems: 8
                       type: array
+                    timeouts:
+                      description: "Timeouts defines the timeouts that can be configured
+                        for an HTTP request. \n Support: Core \n <gateway:experimental>"
+                      properties:
+                        request:
+                          description: "Request specifies an end-to-end timeout for
+                            client HTTP requests, disabled by default. \n Request
+                            timeout values are formatted as 1h/1m/1s/1ms and MUST
+                            BE >= 1ms. \n For example, the following rule will wait
+                            no longer than 2 secs for the client request to complete:
+                            \n ``` rules: - timeouts: request: 2s backendRefs: ...
+                            ``` \n Support: Core"
+                          format: int64
+                          type: integer
```

Not sure how to make it happen, but I think we want something like this?
```suggestion
                          format: duration
                          type: string
```

---

in Contour we did the following: https://github.com/projectcontour/contour/blob/023f6075399213ef7a2c54f6b2853e07e7713c8c/apis/projectcontour/v1/httpproxy.go#L217-L218

the type is string, and we used a kubebuilder validation regex to ensure the right format is passed in

(`time.Duration` devolves down to an int64, the nice `1s` is just string representation)

---

Thanks @sunjayBhatia, that's Interesting. I wonder how the Contour `// +kubebuilder:validation:Pattern=regex` compares to `// +kubebuilder:validation:Format=duration`? I tried to create a field with format set to duration and it seemed to be enforcing time.Duration string format, i.e., it accepted "2s", but didn't accept "hello". I'm not sure where these format values are documented though.

---

@sunjayBhatia thanks for sharing the above. When defining the type as follows:

```go
...
	// +kubebuilder:validation:Type=string
	// +kubebuilder:validation:Format=duration
	Request *time.Duration `json:"request,omitempty"`
...
```

The generated CRD looks like this:

```yaml
...
                    timeouts:
                      description: "Timeouts defines the timeouts that can be configured
                        for an HTTP request. \n Support: Core \n <gateway:experimental>"
                      properties:
                        request:
                          allOf:
                          - format: int64
                          - format: duration
                          description: "Request specifies an end-to-end timeout for
                            client HTTP requests, disabled by default. \n Request
                            timeout values are formatted as 1h/1m/1s/1ms and MUST
                            BE >= 1ms. \n For example, the following rule will wait
                            no longer than 2 secs for the client request to complete:
                            \n ``` rules: - timeouts: request: 2s backendRefs: ...
                            ``` \n Support: Core"
                          type: integer
                      type: object
...
```

The generated `request.allOf` and `request.type: integer` are not what we need. Instead, we need `request.type: string` and `request.format: duration`. I am not familiar with kubebuilder markers and I could not find a clear example anywhere in the official docs https://book.kubebuilder.io/reference/markers/crd-validation.html 

Would you have any idea what I am doing wrong?

---

you might want to try to use the `metav1.Duration` wrapper type rather than `time.Duration` https://github.com/kubernetes/apimachinery/blob/0ef2ec80f7a200acb85cc11636ab3a14a765b6f8/pkg/apis/meta/v1/duration.go#L24-L29

i think you should be able to remove the `validation:Type=string` line as well

---

this worked as expected, thanks @sunjayBhatia